### PR TITLE
feat: enable HD wallets by default

### PR DESCRIPTION
**Pull request by knst**
## Issue being fixed or feature implemented
HD wallets are old-existsing feature, appeared in Dash years ago, but enabling HD wallets is not trivial task that requires multiple steps and command line/rpc calls.
Let's have them enabled by default.

## What was done?
 - HD wallets are enabled by default. Currently behavior `dashd`, `dash-qt` are similar to run with option `-usehd=1`
 - the rpc `upgradewallet` do not let to upgrade from non-HD wallet to HD wallet to don't encourage user use non-crypted wallets (postponed till v21)
 - the initialization of ScriptPubKey is updated to be sure that encypted HD seed is never written on disk (if passphrase is provided)
 - enabled and dashified a script `wallet_upgradewallet.py` which test compatibility between different versions of wallet

## What is not done?
 - wallet tool still does not support passhprase, HD seed can appear on disk
 - there's no dialog that show user a mnemonic phrase and encourage him to make a paper backup
 
Before removing a command line 'usehd' (backport bitcoin#11250) need to make at least one major release for fail-over option (if someone wish to use non-HD wallets only).

## How Has This Been Tested?
Run unit and functional tests.
Enabled new functional test `wallet_upgradewallet.py` that has been backported long time ago but waited this PR to be enabled.

## Breaking Changes
HD wallets are created by default. 

## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -2681,76 +2681,53 @@ static UniValue upgradetohd(const JSONRPCRequest& request)
...
-        // Do not do anything to HD wallets
-        if (pwallet->IsHDEnabled()) {
-            throw JSONRPCError(RPC_WALLET_ERROR, "Cannot upgrade a wallet to HD if it is already upgraded to HD.");
-        }
-
-        if (!pwallet->SetMaxVersion(FEATURE_HD)) {
-            throw JSONRPCError(RPC_WALLET_ERROR, "Cannot downgrade wallet");
+    SecureString secureWalletPassphrase;
+    secureWalletPassphrase.reserve(100);
+    // TODO: get rid of this .c_str() by implementing SecureString::operator=(std::string)
```
Comment (kwvg) [inline src/wallet/rpcwallet.cpp:2690]: _Note: This review comment is not a blocker for this pull request_ 

This TODO has been lingering around for ~12 years in Bitcoin Core (as of this writing) and it seems to be finally removed with https://github.com/bitcoin/bitcoin/pull/27068. Might be worth to backport OOO.

File: src/wallet/rpcwallet.cpp [hunk 2]
```diff
@@ -2681,76 +2681,53 @@ static UniValue upgradetohd(const JSONRPCRequest& request)
...
-
-        // Do not do anything to HD wallets
-        if (pwallet->IsHDEnabled()) {
-            throw JSONRPCError(RPC_WALLET_ERROR, "Cannot upgrade a wallet to HD if it is already upgraded to HD.");
-        }
-
-        if (!pwallet->SetMaxVersion(FEATURE_HD)) {
-            throw JSONRPCError(RPC_WALLET_ERROR, "Cannot downgrade wallet");
+    SecureString secureWalletPassphrase;
+    secureWalletPassphrase.reserve(100);
```
Comment (kwvg) [inline src/wallet/rpcwallet.cpp:2689]: The GUI [reserves](https://github.com/bitcoin/bitcoin/blame/411ba32af21a56efa0a570b6aa8bf8f035410230/src/qt/askpassphrasedialog.cpp#L89-L91) [`MAX_PASSPHRASE_SIZE`](https://github.com/dashpay/dash/blob/ce67e07294c90bcd9a0ffb038bf8aaad0670775e/src/qt/guiconstants.h#L15) for the wallet. 

It's unlikely someone will actually come around with a password >100 characters but since we allow longer passwords in the GUI, we should have RPC behave in the same way.

File: src/wallet/wallettool.cpp [hunk 3]
```diff
@@ -28,9 +28,11 @@ static void WalletCreate(CWallet* wallet_instance)
 
     // generate a new HD seed
     wallet_instance->SetupLegacyScriptPubKeyMan();
-    // NOTE: we do not yet create HD wallets by default
-    // auto spk_man = wallet_instance->GetLegacyScriptPubKeyMan();
-    // spk_man->GenerateNewHDChain("", "");
+    auto spk_man = wallet_instance->GetLegacyScriptPubKeyMan();
+    // NOTE: drop this condition after removing option to create non-HD wallets
+    if (spk_man->IsHDEnabled()) {
+        spk_man->GenerateNewHDChain("", "");
```
Comment (kwvg) [inline src/wallet/wallettool.cpp:34]: _Note: This review comment is not a blocker for this pull request_

Minor nit

```c++
        spk_man->GenerateNewHDChain("" /* secureMnemonic */, "" /* secureMnemonicPassphrase */);
```

File: test/functional/wallet_upgradewallet.py [hunk 4]
```diff
@@ -131,12 +126,17 @@ def run_test(self):
         assert_equal('hdseedid' in wallet.getwalletinfo(), False)
         # calling upgradewallet with explicit version number
         # should return nothing if successful
-        assert_equal(wallet.upgradewallet(169900), "")
+        assert_equal(wallet.upgradewallet(120200), "")
+
+        new_version = wallet.getwalletinfo()["walletversion"]
+        # upgraded wallet should have version 120200
+        #assert_equal(new_version, 120200)
```
Comment (kwvg) [inline test/functional/wallet_upgradewallet.py:133]: Possible TODO for commented out line?

**Review by kwvg — COMMENTED** [review]
LGTM, minor notes

File: src/wallet/wallettool.cpp (hunk 3 above)
Comment (PastaPastaPasta) [reply src/wallet/wallettool.cpp:34]: Need to use correct format at least :)
```
        spk_man->GenerateNewHDChain(/*secureMnemonic=*/"", /*secureMnemonicPassphrase=*/"" );
```

File: src/wallet/wallet.cpp [hunk 5]
```diff
@@ -4897,7 +4897,45 @@ bool CWallet::UpgradeWallet(int version, bilingual_str& error)
         return false;
     }
 
+    // TODO: consider discourage users to skip passphrase for HD wallets for v21
```
[the remark this answers is not in the document]
Comment (knst) [reply src/wallet/wallet.cpp:4900]: An user of dash-core can already have air-gapped device, or it can be encrypted disk, virtual machine, etc.

Forcing user to use passphrase for light client maybe a good idea, but for case of cold wallet it may lead to lost funds if user will forget passphrase after creating one "just do not bother me, here's a simple password" that he may not remember because uses device once in awhile. so, users should be discourage, but not forced IMO

File: src/wallet/wallettool.cpp (hunk 3 above)
Comment (kwvg) [reply src/wallet/wallettool.cpp:34]: To be fair, we seem to use [every](https://github.com/dashpay/dash/blob/ce67e07294c90bcd9a0ffb038bf8aaad0670775e/src/wallet/rpcdump.cpp#L255) [possible](https://github.com/dashpay/dash/blob/ce67e07294c90bcd9a0ffb038bf8aaad0670775e/src/test/addrman_tests.cpp#L421) [variant](https://github.com/dashpay/dash/blob/ce67e07294c90bcd9a0ffb038bf8aaad0670775e/src/net_processing.cpp#L2733-L2735) though the last one is what upstream seems to prefer even if I find it unsightly. Was only suggesting that there be an inline comment and wrote one for convenience...

File: src/wallet/wallet.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (kwvg) [reply src/wallet/wallet.cpp:4900]: There is definitely a risk of forgetting a password but that's inherent to any form of encryption, that be using Cryptomator, BitLocker or FileVault, all three urge you to have a physical copy of the key for recovery should you forget your password. Since we're dealing with user funds here, assuming a mildly clumsy pair of hands (someone who forgot to delete unencrypted copies of their wallet) instead of a cautious pair of hands (a seasoned user who has encryption schemes for their protected environment in lieu of not using a password for their wallet) would be reducing user funds at risk, IMO.

[the remark this answers is not in the document]
Comment (knst) [reply src/wallet/wallet.cpp:4900]: > someone who forgot to delete unencrypted copies of their wallet

you may delete it, but they are still written on disk, until you re-write that sector of disk (what is close to impossible for ssd)

[the remark this answers is not in the document]
Comment (kwvg) [reply src/wallet/wallet.cpp:4900]: > I doubt most users delete their unencrypted wallet file by first overwriting its content with random data and then deleting it ("shredding")

I did mention most people don't "shred" the file and opt to simply delete it but my example of a clumsy user would be one who didn't delete the file at all, because even if the file isn't "shredded", a deleted file overwritten with new data in the now freed up region overwriting (maybe parts of it) is still more frustrating to recover than one that's intact and forgotten about, which can just be copied over.

> what is close to impossible for ssd

Writing repeated patterns is something SSDs try to "optimize away" and there is some truth to wear optimization complicating things but overwriting the file with random data and then deleting it would frustrate recovery attempts much more significantly than plain deletion.

---

The bottom line is, I think security policies should be stricter with opt-out ability, rather than having softer policies and an opt-in to stricter policies.

[the remark this answers is not in the document]
Comment (knst) [reply src/wallet/wallet.cpp:4900]: > Writing repeated patterns is something SSDs try to "optimize away" and there is some truth to wear optimization complicating things but overwriting the file with random data and then deleting it would frustrate recovery attempts much more significantly than plain deletion.

not exactly; I meant that writing on SSD same logical block happen everytime to different physical blocks; and when you trying to `shred` something 100 times - it would be 100 different blocks written, but not your original one due to wear optimization; logical blocks no need to be mapped next-to-each other as on hdd, they can be complely different physical blocks each recording.

[the remark this answers is not in the document]
Comment (kwvg) [reply src/wallet/wallet.cpp:4900]: Then I seem to be mistaken about how "shredding" works on SSDs in that it doesn't work at all due to fundamental decoupling of logical and physical blocks with wear levelling taking the wheel. There are methods that claim to do the job but I can't substantiate them, the only "proper" means seem to be using ATA Secure Erase but that applies to the whole drive, not just the file.

Though this seems to point in the direction that we should avoid writing unencrypted data at all since overwriting the file _alone_ isn't feasible.

File: src/wallet/rpcwallet.cpp (hunk 1 above)
Comment (knst) [reply src/wallet/rpcwallet.cpp:2690]: I will do it in the next PRs after #5807 (this one) and #5854 (HD wallet refactorings).

Alongs #27068, I'd like to introduce std::optional in hdchain and couple more changes

File: src/wallet/rpcwallet.cpp (hunk 2 above)
Comment (knst) [reply src/wallet/rpcwallet.cpp:2689]: This call `reserve` has been introduced before introducing SecureString (which use custom allocator).
That call is needed to be sure that string is not accidentally re-allocated and it would be impossible to wipe a data by rewriting bytes.

This call is just for legacy and completely useless now. I'd better to consider removing `reserve` completely rather than keep it here.

btw, that's not new code, I just moved it a bit; I will consider removing `reserve` at all, not planning to refactor it in this PR

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
LGTM overall, check comments below for 2 small issues (fixes: cda12bddb6)

**Review by UdjinM6 — APPROVED** [review]
light ACK

**Review by ogabrielides — APPROVED** [review]
LGTM in general

**Review by PastaPastaPasta — CHANGES_REQUESTED** [review]
Tested; looks good. But needs release notes :)

File: doc/release-notes-5807.md [hunk 6]
```diff
@@ -0,0 +1,8 @@
+# Notable changes
+
+## HD Wallets Enabled by Default
+
+In this release, we are taking a significant step towards enhancing the Dash wallet's usability by enabling Hierarchical Deterministic (HD) wallets by default. This change aligns the behavior of `dashd` and `dash-qt` with the previously optional `-usehd=1` flag, making HD wallets the standard for all users.
+Dash Core still can be enforced by creating non-HD wallet by run with `-usehd=0` flag but this feature will be removed in next releases.
```
Comment (UdjinM6) [inline doc/release-notes-5807.md:6]: feels like a duplicate
```suggestion
```

Comment (knst) [reply doc/release-notes-5807.md:6]: @UdjinM6 
maybe, 

```
Command line argument `-usehd` is to remove in next major release
```
?

Comment (UdjinM6) [reply doc/release-notes-5807.md:6]: I mean you already have

>However, users should be aware that this option is intended for legacy support and will be removed in future releases.

below, so why saying it again?

**Comment by UdjinM6** [comment]
LGTM but I'll wait for @thephez :)

**Review by thephez — APPROVED** [review]
Release notes look okay to me :+1:

**Review by ogabrielides — APPROVED** [review]
LGTM in general, good job +1

**Review by PastaPastaPasta — APPROVED** [review]
utACK
